frame/transaction-payment/asset-conversion-tx-payment/src/payment.rs

```diff
@@ -0,0 +1,183 @@
+	/// amount in native currency used to pay the fee.
+	fn withdraw_fee(
+		who: &T::AccountId,
+		call: &T::RuntimeCall,
+		info: &DispatchInfoOf<T::RuntimeCall>,
+		asset_id: Self::AssetId,
+		fee: BalanceOf<T>,
+		tip: BalanceOf<T>,
+	) -> Result<
+		(LiquidityInfoOf<T>, Self::LiquidityInfo, AssetBalanceOf<T>),
+		TransactionValidityError,
+	> {
+		// convert the asset into native currency
+		let ed = C::minimum_balance();
+		let native_asset_required =
+			if C::balance(&who) >= ed.saturating_add(fee.into()) { fee } else { fee + ed.into() };
+
+		let asset_consumed = CON::swap_tokens_for_exact_native(
+			who.clone(),
+			asset_id,
+			native_asset_required,
+			None,
+			who.clone(),
+			true,
+		)
+		.map_err(|_| TransactionValidityError::from(InvalidTransaction::Payment))?;
+
+		ensure!(asset_consumed > Zero::zero(), InvalidTransaction::Payment);
+
+		// charge the fee in native currency
+		<T::OnChargeTransaction>::withdraw_fee(who, call, info, fee, tip)
+			.map(|r| (r, native_asset_required, asset_consumed))
+	}
+
+	/// Correct the fee and swap the refund back to asset.
+	///
+	/// Note: The `corrected_fee` already includes the `tip`.
+	/// Note: Is the ED wasn't needed, the `received_exchanged` will be equal to `fee_paid`, or
+	/// `fee_paid` + ed otherwise
+	fn correct_and_deposit_fee(
+		who: &T::AccountId,
+		dispatch_info: &DispatchInfoOf<T::RuntimeCall>,
+		post_info: &PostDispatchInfoOf<T::RuntimeCall>,
+		corrected_fee: BalanceOf<T>,
+		tip: BalanceOf<T>,
+		fee_paid: LiquidityInfoOf<T>,
+		received_exchanged: Self::LiquidityInfo,
+		asset_id: Self::AssetId,
+		initial_asset_consumed: AssetBalanceOf<T>,
+	) -> Result<AssetBalanceOf<T>, TransactionValidityError> {
+		// Refund the native asset to the account that paid the fees (`who`).
+		// The `who` account will receive the "fee_paid - corrected_fee" refund.
+		<T::OnChargeTransaction>::correct_and_deposit_fee(
+			who,
+			dispatch_info,
+			post_info,
+			corrected_fee,
+			tip,
+			fee_paid,
+		)?;
+
+		// calculate the refund in native asset, to swap back to the desired `asset_id`
+		let swap_back = received_exchanged.saturating_sub(corrected_fee);
+		let mut asset_refund = Zero::zero();
+		if !swap_back.is_zero() {
+			// If this fails, the account might have dropped below the existential balance or there
+			// is not enough liquidity left in the pool. In that case we don't throw an error and
+			// the account will keep the native currency.
+			asset_refund = CON::swap_exact_native_for_tokens(
+				who.clone(), // we already deposited the native to `who`
+				asset_id,    // we want asset_id back
+				swap_back,   // amount of the native asset to convert to `asset_id`
+				None,        // no minimum amount back
+				who.clone(), // we will refund to `who`
+				false,       // no need to keep alive
+			)
+			.map_err(|_| TransactionValidityError::from(InvalidTransaction::Payment))?;
```

Comment:
It does throw an error though (which it probably shouldn't). IMO if this fails it should set `asset_refund = 0` and perhaps emit a different event, like
```rust
AssetRefundFailed {
    native_amount_kept: swap_back,
}
```
Now in the `AssetTxFeePaid` event you will see a high amount for `actual_fee` in `asset_id`, and this extra event tells you that some native asset was also left on the account.

Reply:
Done